pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh

```diff
@@ -0,0 +1,68 @@
+# shellcheck shell=bash
+
+yarnInstallHook() {
+    echo "Executing yarnInstallHook"
+
+    runHook preInstall
+
+    local -r packageName="$(@jq@ --raw-output '.name' package.json)"
+    local -r packageOut="$out/lib/node_modules/$packageName"
+
+    # TODO: figure out a way to avoid redundant compress/decompress steps
+    yarn pack --filename "$TMPDIR/$packageName.dist.tgz" \
+        --ignore-engines \
+        --ignore-platform \
+        --ignore-scripts \
+        --non-interactive \
+        --offline
+
+    mkdir -p "$packageOut"
+    tar -xzf "$TMPDIR/$packageName.dist.tgz" \
+        -C "$packageOut" \
+        --strip-components 1 \
+        package/
```

Comment:
Yea indeed this is a bit weird.. Isn't that doing what `cp -r node_modules` does below?

Reply:
No, this is installing only the package's files, `cp -r node_modules` is installing only the package's dependencies.

Here's an example of the contents of the dist file for `blade-formatter`, note the absence of `node_modules`:

<details><summary>Details</summary>
<pre>
<code>
$ tar tf blade-formatter.dist.tgz
package
package/CHANGELOG.md
package/LICENSE
package/README.md
package/bin
package/dist
package/package.json
package/src
package/syntaxes
package/bin/blade-formatter.cjs
package/bin/blade-formatter.js
package/dist/types
package/src/cli.ts
package/src/comment.ts
package/src/constants.ts
package/src/errors
package/src/formatter.ts
package/src/indent.ts
package/src/main.ts
package/src/regex.ts
package/src/runtimeConfig.ts
package/src/space.ts
package/src/util.ts
package/src/vsctm.ts
package/syntaxes/blade.tmLanguage.json
package/dist/types/comment.d.ts
package/dist/types/constants.d.ts
package/dist/types/errors
package/dist/types/formatter.d.ts
package/dist/types/indent.d.ts
package/dist/types/main.d.ts
package/dist/types/regex.d.ts
package/dist/types/runtimeConfig.d.ts
package/dist/types/space.d.ts
package/dist/types/util.d.ts
package/dist/types/vsctm.d.ts
package/src/errors/formatError.ts
package/dist/types/errors/formatError.d.ts
</code>
</pre>
</details> 

This retains the same functionality of [these lines](https://github.com/NixOS/nixpkgs/blob/5154507a1b18ae480339201b8f55b145fd52a0f0/pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh#L10-L15) from npmInstallHook, but translated to `yarn`.

The problem is that `yarn pack` lacks the `--dry-run` flag, in fact the only flag that it supports is `--filename`: https://classic.yarnpkg.com/en/docs/cli/pack

The `--dry-run` flag does exist, but only in yarn v3+: https://v3.yarnpkg.com/cli/pack

It would be nice to tackle is v3 support too some time: https://github.com/NixOS/nixpkgs/pull/337545/files#diff-1fc5bfea9a2e2f459eb9ad2823a7f1048e5c71e444df55cc774ea114ea5dec23

Reply:
> It would be nice to tackle is v3 support too some time: [#337545 (files)](https://github.com/NixOS/nixpkgs/pull/337545/files#diff-1fc5bfea9a2e2f459eb9ad2823a7f1048e5c71e444df55cc774ea114ea5dec23)

What exactly should I be looking at there?

Reply:
People are resorting to making their own FOD in order to use yarn V2, since fetchYarnDeps does not support it.

Reply:
> People are resorting to making their own FOD in order to use yarn V2, since fetchYarnDeps does not support it.

And there's also yarn v3 available?

Reply:
v1 is what we support in nixpkgs (https://github.com/yarnpkg/yarn),

All other versions are in https://github.com/yarnpkg/berry

v2 is when changes were made to the lockfile format,
v3 is the current stable version,
v4 is the next version, not yet stable, AFAIK.

Reply:
Mentioning https://github.com/NixOS/nixpkgs/issues/254369 here, and marking conversation as resolved.